Automized Server tests

CanBerker opened the pull request:
Firstly, I have moved spotless checks before all other tests so formatting issues can be detected first without wait.

Secondly, I have implemented a bash script with automized server tests. This works on my local machine. ~However, since the workflow does not install spark-submit to the test machine. The server initialization fails.~ I wanted to get @ghislainfourny's feedback on how to proceed.

Edit: I have updated the script to install Spark during testing

Fixes #664

CanBerker committed:
Implement automized server tests

CanBerker committed:
Install Spark before running server test

ghislainfourny commented:
This is wonderful, thank you!

Could we please add a test that checks for a multiline query in the request body? There was a bug we had to fix to make it work with the Rumble magic in Jupyter notebooks.

After this, we could merge. This is a perfect bootstrap of server tests.

I think installing Spark in this way is good and I am happy with this. I guess the maven test command might create a jar where it is included or something, which explains why it is not needed there.

ghislainfourny commented on .github/workflows/maven.yml:
```diff
@@ -34,8 +36,8 @@ jobs:
-    - name: Spotless check
-      run: mvn spotless:check
+    - name: Install Spark and run server tests
+      run: mkdir -p /opt; wget -q -O /opt/spark.tgz https://downloads.apache.org/spark/spark-2.4.5/spark-2.4.5-bin-hadoop2.7.tgz ; tar -xzf /opt/spark.tgz -C /opt/ ; rm /opt/spark.tgz ; export SPARK_HOME=/opt/spark-2.4.5-bin-hadoop2.7 ; export PATH=$PATH:/opt/spark-2.4.5-bin-hadoop2.7/bin; bash server-test-script.sh
```
Do you think we could split this into several intermediate step for ease of read?

CanBerker replied:
For readability I have initially aimed to use multi-line commands which are supposedly supported by workflow .yml. However, when tested, it would only run the first command and skip the rest. So I have opted for chaining them via semi-colons.

The reason I have put the entire Spark installation and testing in a single step is to make it clear that this is the only step that uses the Spark installation. 

I will nonetheless, go over the alternatives once more to make this more readable. I would be glad to hear any preferences you may have(no matter weak or strong) on how to structure this.

ghislainfourny commented:
We'll have to keep in mind to bump the version here as well (the minor update to 1.6.1 broke the tests but should be an easy fix).

CanBerker commented:
> Could we please add a test that checks for a multiline query in the request body? There was a bug we had to fix to make it work with the Rumble magic in Jupyter notebooks.

If you could share the test that failed earlier I could add it as our test. Otherwise I'm assuming any test containing a new line character "\n"  in the request should do. Is this correct?

CanBerker committed:
Remove version dependency from server tests

CanBerker committed:
Remove stacktrace from test5

CanBerker commented:
> We'll have to keep in mind to bump the version here as well (the minor update to 1.6.1 broke the tests but should be an easy fix).

The version changes should not be a problem anymore with my recent commits.

CanBerker committed:
Add multiline tests

CanBerker commented:
@ghislainfourny the newline added with "\n" breaks the query for some reason. Could you provide any insights or could you share the query that you intended to run?

Commit:
improve sequencelookup df and optimize rdd case

Commit:
fix integer predicate to fallback even in case native execution is activated

Commit:
sequencelookup with iteration instead of materialization

Commit:
optimize predicate of type [position() eq/lt/gt x] to lookup/subsequence

Commit:
fix executionmode for indexlookup

Commit:
Update .gitlab-ci.yml file

Commit:
Update .gitlab-ci.yml file

Commit:
Update .gitlab-ci.yml file on my master to check its tests

Commit:
deactivate optimization for now, still needs executionmodevisitor anyways

Commit:
Update .gitlab-ci.yml environment

Commit:
trying something

Commit:
try maven image

Commit:
maven spotless

Commit:
sequencelookup fix dynamic context

Commit:
sequencelookup fix by properly closing iterator

Commit:
try actions from david

mschoeb committed:
position() eq optimization and cleanup

mschoeb committed:
Merge branch 'master3.5-mschoeb' into mschoeb_firstwork

# Conflicts:
#	.gitlab-ci.yml

mschoeb committed:
update delta spark version

mschoeb committed:
scala 2.13 upgrade

mschoeb committed:
better sequencelookup iteration and todos from meeting

mschoeb committed:
fix parallelize for dataframes

mschoeb committed:
maven spotless

mschoeb committed:
optimize float in predicate index and work according to spec in case of float with fractions

mschoeb committed:
update test like on github

mschoeb committed:
fix test typo

Commit:
Merge branch 'mschoeb_firstwork' into 'master3.5-mschoeb'

Mschoeb first features

See merge request gfourny/rumble!84

mschoeb committed:
Merge remote-tracking branch 'origin/xml-support' into mschoeb-xml

# Conflicts:
#	.gitlab-ci.yml
#	src/main/java/org/rumbledb/items/parsing/ItemParser.java
#	src/main/java/org/rumbledb/parser/Jsoniq.g4
#	src/main/java/org/rumbledb/parser/Jsoniq.interp
#	src/main/java/org/rumbledb/parser/Jsoniq.tokens
#	src/main/java/org/rumbledb/parser/JsoniqLexer.interp
#	src/main/java/org/rumbledb/parser/JsoniqLexer.java
#	src/main/java/org/rumbledb/parser/JsoniqLexer.tokens
#	src/main/java/org/rumbledb/parser/JsoniqParser.java

mschoeb committed:
fix isAnnotation and rerun antlr

mschoeb committed:
remove old (if i understood it correctly) %an: annotations
revert test of describe for now

mschoeb committed:
first nonworking version of xml-files

mschoeb committed:
update kryo write functions to support node

mschoeb committed:
fix serialization by doing some odd sideeffecting call

mschoeb committed:
revert some notations

mschoeb committed:
some cleanup

mschoeb committed:
Merge branch 'master' into ghislainfourny-patch-1

ghislainfourny committed:
Merge branch 'master' into serverTests

ghislainfourny committed:
Update maven.yml

ghislainfourny commented:
@CanBerker this PR was somehow forgotten, it would be nice if we merge this if all tests pass.

mschoeb committed:
Merge pull request #1263 from RumbleDB/ghislainfourny-patch-1

Update install.md

mschoeb committed:
fix text keyword and regenerate grammar

mschoeb committed:
mvn spotless

mschoeb committed:
update some more kryo problems

mschoeb committed:
make tests serializable

mschoeb committed:
first version of pathexpriterator that can take rdds

mschoeb committed:
small update to first flawed version

mschoeb committed:
basic tests for xml-files function

mschoeb committed:
refactor flat pathexpr into nested tree of slashexpr

mschoeb committed:
remove path expression implementations

mschoeb committed:
remove unnessecary addition of applyinpath

mschoeb committed:
order xml nodes if RDD

mschoeb committed:
temporary workaround for slashexpr rdd if right side is not step

mschoeb committed:
add file path to key for sorting and uniqueness

mschoeb committed:
fix wrong indexlookup test and update slashexpr implementation accordingly

mschoeb committed:
minor cleanup and comments

mschoeb committed:
deactivate faulty test for now

mschoeb committed:
fix uniqueness and order across documents

Commit:
Merge branch 'mschoeb-xml' into 'master3.5-mschoeb'

Update and extend XML support

See merge request gfourny/rumble!85

mschoeb committed:
remove wrong occurence of warning of old parameters

mschoeb committed:
update xml-doc to doc

mschoeb committed:
json-lines to replace json-file, deprecate json-file
same for structured-json-lines

mschoeb committed:
more renaming from json-file to json-lines

mschoeb committed:
discard most test output for cleaner logs

mschoeb committed:
try logging junit in gitlab

mschoeb committed:
try again

mschoeb committed:
reporting for all tests and spotless

mschoeb committed:
fix format

mschoeb committed:
fix maven warnings

mschoeb committed:
spotless

mschoeb committed:
disable describe test the proper way

mschoeb committed:
use default 30day expiry

mschoeb committed:
temurin image to compare performance

mschoeb committed:
revert to old container

Commit:
Merge branch 'mschoeb_smallthings' into 'master3.5-mschoeb'

Small changes all over the place

See merge request gfourny/rumble!86

mschoeb committed:
fix boolean value of double NaN should be false
fn-boolean-mixed-args-017
fn-boolean-mixed-args-023

mschoeb committed:
fix bigdecimal compareto instead of equals
K-SeqBooleanFunc-22

mschoeb committed:
fix objects and arrays not having true as effective boolean value, expecting this to maybe break some of our tests
boolean-010
boolean-010b

mschoeb committed:
update two tests

mschoeb committed:
update two tests again

mschoeb committed:
fix string to codepoint on multibyte range
cbcl-codepoints-to-string-015
cbcl-codepoints-to-string-016

mschoeb committed:
fix encode * to %2A
fn-encode-for-uri-8

mschoeb committed:
fix indexof shouldnt find NaN
fn-indexof-mix-args-009
K-SeqIndexOfFunc-10

mschoeb committed:
fix execution mode warning
ported from github

mschoeb committed:
fix broken nonfloat indexof cases

mschoeb committed:
support ' in jsoniq test

mschoeb committed:
revert booleanvalue change, will be in 3.1 later

mschoeb committed:
skellington for effectivebooleanvalue

mschoeb committed:
concat needs at least 2 arguments

Commit:
Merge branch 'mschoeb-tests-bugfixing' into 'master3.5-mschoeb'

Bugfixes found through testsuite

See merge request gfourny/rumble!87

mschoeb committed:
add cli option for jsoniq version

mschoeb committed:
fix cli docs formatting

mschoeb committed:
try mvn container for pipeline

mschoeb committed:
install ant

mschoeb committed:
dont rebuild parser

mschoeb committed:
remove generated files from repo

mschoeb committed:
update gitignore to not track generated files

mschoeb committed:
spotless and updated documentation without antlr build

mschoeb committed:
test trigger

mschoeb committed:
pass tested branch

mschoeb committed:
pass tested branch

mschoeb committed:
run all tests at start, run testsuite only if all tests pass

Commit:
Merge branch 'mschoeb-downstream-pipeline' into 'master3.5-mschoeb'

Run testsuite on commit in background

See merge request gfourny/rumble!89

mschoeb committed:
fix mod with bigInteger

mschoeb committed:
allow item promotable to string as key in objectconstructor

mschoeb committed:
concat shouldnt take sequences, only single atomics

mschoeb committed:
keep history for shell

mschoeb committed:
add compare function

mschoeb committed:
fix typo

mschoeb committed:
trace with arity 1

mschoeb committed:
hopefull fix numeric

mschoeb committed:
fix numeric from boolean

mschoeb committed:
fix default empty error thrown

mschoeb committed:
spotless

mschoeb committed:
change base64 regex to hopefully handle whitespace correctly

mschoeb committed:
fix base64 == case

mschoeb committed:
remove debug output from reports that get too large

mschoeb committed:
rename atomic to anyAtomicType with backwards compatibility

mschoeb committed:
fix some edgecases in math functions

mschoeb committed:
effective boolean value of nodes

mschoeb committed:
fix effective boolean value

mschoeb committed:
fix sum when zeroElement is non-empty

mschoeb committed:
spotless

mschoeb committed:
wrap matches with correct exception

mschoeb committed:
remove unnecessary if in resolveuri

mschoeb committed:
require quoted keys in jsoniq3.1
THIS REQUIRES ABBREVSTEPS TO WORK WHICH DONT YET

mschoeb committed:
more atomic to anyAtomicType

mschoeb committed:
change more tests to anyAtomicType

mschoeb committed:
update effectivebooleanval tests

mschoeb committed:
fix more tests and add backwards compatibility testing

mschoeb committed:
fix test and prints

mschoeb committed:
revert incorrect fix for one test
add option to specify jsoniq version in query

mschoeb committed:
specify jsoniq version in tests that specificially need it

mschoeb committed:
switch default to jsoniq 1.0

mschoeb committed:
fix test line missmatch

mschoeb committed:
bump to java 11

mschoeb committed:
bump some dependencies

mschoeb committed:
Merge branch 'mschoeb_jsoniq31' into 3.1-require-quoted-keys

Commit:
Merge branch 'bugfixing-8.1' into 'master3.5-mschoeb'

remove unnecessary if in resolveuri

See merge request gfourny/rumble!90

mschoeb committed:
change argumentlist to require commas between args

mschoeb committed:
fix tests, remove some numpy scripting blocks that caused issues wiht abbreviated steps

mschoeb committed:
add atomic to anyAtomicType renaming to more places

Commit:
Merge branch 'mschoeb-version-bumps' into 'master3.5-mschoeb'

upgrade java version and some dependencies

See merge request gfourny/rumble!91

mschoeb committed:
proper divide by zero exception in duration

mschoeb committed:
unary error if non numeric

mschoeb committed:
fix idiv float edgecase

mschoeb committed:
more precision in decimal multiplication
[op/numeric-divide.xml] op-numeric-dividedec2args-4

mschoeb committed:
nan ne something is always true
[op/numeric-equal.xml] K-NumericEqual-32

mschoeb committed:
duration * infinity is overflow
[op/multiply-yearMonthDuration.xml] K-YearMonthDurationMultiply-6

mschoeb committed:
fix tests with higher precision

mschoeb committed:
Add FODT0001 and fix some FODT0002 cases

mschoeb committed:
fix durations and datetime exceptions

mschoeb committed:
try 2 fix durations and datetime exceptions

mschoeb committed:
add FOAR0002

mschoeb committed:
spotless

mschoeb committed:
fix some more cases

mschoeb committed:
fix imult with float cases

mschoeb committed:
fix durationerror instead of datetime error in duration

mschoeb committed:
fix divide by zero error higher precedence over inf or nan

mschoeb committed:
fix getCurrentDateTime not available in children

mschoeb committed:
spotless

mschoeb committed:
fix position and size missing from inner focus in simplemap

Commit:
Merge branch '3.1-require-quoted-keys' into 'mschoeb_jsoniq31'

require quoted keys in jsoniq3.1

See merge request gfourny/rumble!92

mschoeb committed:
Merge branch 'master3.5-mschoeb' into mschoeb_jsoniq31

mschoeb committed:
fix subsequence edgecase with negative startindex

mschoeb committed:
remove missing this. warning

mschoeb committed:
fix step setting context outside

mschoeb committed:
fix bug I introduced with subsequence without specified length

mschoeb committed:
start implementing separate xml tests

mschoeb committed:
add tests for unquoted keys and test that old things throw correct errors in jsoniq3.1

Commit:
Merge branch 'mschoeb_jsoniq31' into 'master3.5-mschoeb'

Update to JSONiq 3.1 and abbrevstep

See merge request gfourny/rumble!88

mschoeb committed:
separate xml tests

mschoeb committed:
fix unqualified access warning

mschoeb committed:
add xml tests to pipeline

Commit:
Merge branch 'update-tests' into 'master3.5-mschoeb'

Move XML tests separate

See merge request gfourny/rumble!95

mschoeb committed:
change JNDY0003 to XQDY0137

mschoeb committed:
initial structure 1 to 1

mschoeb committed:
rename context item, module decl, remove true/false/null literals

mschoeb committed:
object and array constructors

mschoeb committed:
initial commit for removing w3c.node in java and doing our implementation instead

mschoeb committed:
spotless

mschoeb committed:
do equality with xmldocumentposition

mschoeb committed:
atomization with materialization for now, data() function

mschoeb committed:
cleanup atomization naming and use correct string-values and typed-values for nodes

mschoeb committed:
streaming and spotless

Commit:
Merge branch 'replace-node' into 'master3.5-mschoeb'

Rework nodes and add atomization

See merge request gfourny/rumble!96

mschoeb committed:
Merge branch 'master3.5-mschoeb' into fork-xquery

mschoeb committed:
use getStringValue in nodes

mschoeb committed:
remove empty step

mschoeb committed:
atomization in cast

mschoeb committed:
update exception code

mschoeb committed:
fix map allow whitespace

mschoeb committed:
remove arrayunboxing and arraylookup

mschoeb committed:
xquery lookup as clone of object lookup

mschoeb committed:
xquery lookup first version done

mschoeb committed:
Merge branch 'master3.5-mschoeb' into fork-xquery

mschoeb committed:
fix xml-files() in FLWOR,
use kryo.setReference(true) for now since we have circular refs
add necessary registration and empty constructors

mschoeb committed:
cosmetics

mschoeb committed:
remove isStaticallyCastable and move atomization into castItemToType

mschoeb committed:
isStaticallyCastable inside castItemToType

mschoeb committed:
fix RumbleML to use serialization instead of cast on Arrays and Objects

mschoeb committed:
formatting of error messges

mschoeb committed:
spotless

Commit:
Merge branch 'mschoeb-cast' into 'master3.5-mschoeb'

Atomization in cast

See merge request gfourny/rumble!99

mschoeb committed:
XPDY when context is missing

mschoeb committed:
add skeleton for kindtest for now, very unfinished

mschoeb committed:
add wildcard to lookup

mschoeb committed:
remove not expression

mschoeb committed:
rename lookup to postfixlookup

mschoeb committed:
implement unary lookup

mschoeb committed:
readd varref to lookup

mschoeb committed:
spotless

Commit:
Merge branch 'fork-xquery' into 'master3.5-mschoeb'

New XQuery parser forked from jsoniq parser but with the jsoniq parts removed

See merge request gfourny/rumble!97

mschoeb committed:
revert maven version

mschoeb committed:
update documentation

mschoeb committed:
--language to choose between jsoniq10, jsoniq31 and xquery31

mschoeb committed:
some renaming in nodes

mschoeb committed:
xquery parser like jsoniq parser with moduleAndThisIsIt

mschoeb committed:
use distinct() in slashexpression

mschoeb committed:
Merge branch 'master' into master3.5-mschoeb

# Conflicts:
#	docs/install.md
#	pom.xml
#	src/main/java/org/rumbledb/cli/Main.java
#	src/main/java/org/rumbledb/context/BuiltinFunctionCatalogue.java

mschoeb committed:
spotless

mschoeb committed:
add .isNode() to item

mschoeb committed:
fix steps resulting in non-nodes, add XTPY0018

mschoeb committed:
more tests for xml steps in jsoniq

mschoeb committed:
change --language to --default-language

mschoeb committed:
add FOAY0001

mschoeb committed:
add tests for xquery parser features

mschoeb committed:
fix error and remove some unimplemented node related methods

mschoeb committed:
move to 1.23.0 Mountain ash

mschoeb committed:
Merge remote-tracking branch 'origin/master3.5-mschoeb' into master3.5-mschoeb

mschoeb committed:
remove old parser antlr config

mschoeb committed:
upload-artifacts v4, specify files to be skipped by spotless

mschoeb committed:
add new XML and XQuery tests to github workflow

ghislainfourny committed:
Merge pull request #1294 from RumbleDB/master3.5-mschoeb

MSc thesis mschoeb

Byczax committed:
Fix CI/CD to run tests in parralel

resolves #1296

Byczax committed:
Update maven.yml

ghislainfourny committed:
Merge pull request #1297 from Byczax/patch-1

Fix CI/CD to run tests in parralel

ghislainfourny committed:
Update maven.yml

ghislainfourny committed:
Merge pull request #1303 from RumbleDB/merge-ci

Merge ci

Byczax committed:
add bug_report issue template

Byczax committed:
add feature and enchantment template

Byczax committed:
Update .github/ISSUE_TEMPLATE/bug_report.yaml

Co-authored-by: Marco Schöb <marcoschoeb@gmail.com>

Byczax committed:
Update .github/ISSUE_TEMPLATE/bug_report.yaml

Co-authored-by: Marco Schöb <marcoschoeb@gmail.com>

Byczax committed:
Rename enchantment.yaml to enhancement.yaml

ghislainfourny committed:
Merge pull request #1304 from RumbleDB/add-issue-templates

Add issue templates

Byczax committed:
Integrate test suite into new test file

Byczax committed:
Merge remote-tracking branch 'origin/serverTests' into serverTests

# Conflicts:
#	.github/workflows/maven.yml

Byczax commented:
Hello @CanBerker
Is it possible for you to check and fix those test suites that you proposed?
I tried my best to integrate it as CI/CD, but I am worried that some things can be outdated, and I can remove some functionalities that you wanted to preserve.
